R/deploy.R

```diff
@@ -501,55 +501,80 @@ get_vanity_url <- function(content) {
   return(van$path)
 }
 
-#' Swap the Vanity URL
+#' Swap Vanity URLs
 #'
-#' Swaps the Vanity URLs between two pieces of content
+#' Swap the vanity URLs of two pieces of content.
 #'
-#' @param from_content A Content object
-#' @param to_content A Content object
+#' @param content_a A Content object
+#' @param content_b A Content object
+#'
+#' @returns A list of the new vanity URLs for `content_a` and `content_b`
 #'
 #' @family content functions
 #' @export
-swap_vanity_url <- function(from_content, to_content) {
-  warn_experimental("swap_vanity_url")
-  scoped_experimental_silence()
+swap_vanity_urls <- function(content_a, content_b) {
   # TODO: Add prompt if in an interactive session
   # TODO: Add pretty print output of what is happening
-  # TODO: Test error cases super thoroughly!!
-  # TODO: Do a "dry run" of sorts...? Check privileges... etc...
-  # TODO: Do the changes within a TryCatch so we can undo...?
-  # TODO: Need a way to "unset" a vanity URL
-
-  from_vanity <- get_vanity_url(from_content)
-  to_vanity <- get_vanity_url(to_content)
 
-  if (is.null(from_vanity) && is.null(to_vanity)) {
-    warning("Neither content has a Vanity URL. Exiting")
-  } else {
-    # swapping vanity URLs
-    tmp_vanity <- paste0("vanity-url-swap-", create_random_name(length = 50))
+  validate_R6_class(content_a, "Content")
+  validate_R6_class(content_b, "Content")
 
-    if (!is.null(from_vanity)) {
-      set_vanity_url(from_content, tmp_vanity)
-    } else {
-      set_vanity_url(to_content, tmp_vanity)
-    }
+  vanity_a <- get_vanity_url(content_a)
+  vanity_b <- get_vanity_url(content_b)
 
-    if (!is.null(from_vanity)) {
-      set_vanity_url(to_content, from_vanity)
+  if (is.null(vanity_a) && is.null(vanity_b)) {
+    warning("Neither content has a vanity URL")
+  } else {
+    tryCatch(
+      delete_vanity_url(content_a),
+      error = function(e) {
+        stop("Unable to modify the vanity URL for content_a: ", e$message, call. = FALSE)
+      }
+    )
+    tryCatch(
+      delete_vanity_url(content_b),
+      error = function(e) {
+        set_vanity_url(content_a, vanity_a)
+        stop("Unable to modify the vanity URL for content_b: ", e$message, call. = FALSE)
+      }
+    )
+    if (!is.null(vanity_a)) {
+      set_vanity_url(content_b, vanity_a)
     }
-    if (!is.null(to_vanity)) {
-      set_vanity_url(from_content, to_vanity)
+    if (!is.null(vanity_b)) {
+      set_vanity_url(content_a, vanity_b)
     }
-
-    from_vanity <- get_vanity_url(from_content)
-    to_vanity <- get_vanity_url(to_content)
+    vanity_a <- get_vanity_url(content_a)
+    vanity_b <- get_vanity_url(content_b)
```

Comment:
Nice, this is a good idea. The server does mutate the submitted vanity url in some cases (i.e., leading '/').